site/content/blog/2019-10-21-localizing-matterpoll/index.md

```diff
@@ -0,0 +1,188 @@
+```go
+// initBundle loads all localization files in i18n into a bundle and return this
+func (p *MatterpollPlugin) initBundle() (*i18n.Bundle, error) {
+	bundle := &i18n.Bundle{DefaultLanguage: language.English}
+	bundle.RegisterUnmarshalFunc("json", json.Unmarshal)
+
+	bundlePath, err := p.API.GetBundlePath()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get bundle path")
+	}
+
+	i18nDir := filepath.Join(bundlePath, "assets", "i18n")
+	files, err := ioutil.ReadDir(i18nDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open i18n directory")
+	}
+
+	for _, file := range files {
+		if !strings.HasPrefix(file.Name(), "active.") {
+			continue
+		}
+
+		if file.Name() == "active.en.json" {
+			continue
+		}
+		_, err = bundle.LoadMessageFile(filepath.Join(i18nDir, file.Name()))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to load message file %s", file.Name())
+		}
+	}
+
+	return bundle, nil
+}
+```
+
+The bundle returned is stored inside the `MatterpollPlugin` struct. We call `initBundle` once in [`OnActivate`](https://developers.mattermost.com/extend/plugins/server/reference/#Hooks.OnActivate).
+
+A `Localizer` is ephemeral and needs to be created every time a user interacts with the plugin, for example, when creating a poll via `/poll` or voting by pressing an [Interactive Message Button](https://docs.mattermost.com/developer/interactive-messages.html). In order to create a `Localizer`  we need to fetch the user's `Locale` setting. Conveniently, it's part of the [`User`](https://godoc.org/github.com/mattermost/mattermost-server/model#User) struct. We wrote a short helper function to create a `Localizer` for a specific user:
+
+```go
+// getUserLocalizer returns a localizer that localizes in the users locale
+func (p *MatterpollPlugin) getUserLocalizer(userID string) *i18n.Localizer {
+	user, err := p.API.GetUser(userID)
+	if err != nil {
+		p.API.LogWarn("Failed get user's locale", "error", err.Error())
+		return p.getServerLocalizer()
+	}
+
+	return i18n.NewLocalizer(p.bundle, user.Locale)
+}
+```
+
+Not every part of Matterpoll is localizable on a user level. Take polls for example.
+
+{{< figure src="poll-1.png" alt="Poll 1">}}
+
+While the response when clicking an option should be localized with the user's local, the "Total Votes" text cannot. Posts are shown in the same way for every user, hence we have to use the same localization for every user. Admins can configure a [Default Client Language](https://docs.mattermost.com/administration/config-settings.html#default-client-language) in the System Console that is used for newly-created users and pages where the user hasn’t logged in. We decided to use this local for translating these kind of strings and build a helper for this:
```

Comment:
Should this be "local" (i.e., local machine) or "locale" (i.e., location/area)?

Reply:
Good catch. It should be `locale` :+1: